torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -0,0 +1,50 @@
+from typing import Dict, List
+
+import torch
+import torch.nn as nn
+
+from torch.distributed.fsdp._common_utils import _named_parameters_with_duplicates
+
+from ._fsdp_common import FSDPMeshInfo, ParamModuleInfo
+from ._fsdp_param import FSDPParam
+
+
+class FSDPParamGroup:
+    """This class represents a parameter group to communicate together."""
+
+    def __init__(
+        self,
+        params: List[nn.Parameter],
+        module: nn.Module,
+        mesh_info: FSDPMeshInfo,
+        device: torch.device,
+    ):
+        self.module = module  # permit ref cycle because 1:1 lifetime
+        param_module_infos = _get_param_module_infos(params, module)
+        self.fsdp_params = [
+            FSDPParam(param, module_info, mesh_info, device)
+            for param, module_info in zip(params, param_module_infos)
+        ]
+        self.mesh_info = mesh_info
+        self.device = device
+
+
+def _get_param_module_infos(
+    params: List[nn.Parameter], module: nn.Module
+) -> List[ParamModuleInfo]:
+    params_set = set(params)
+    param_to_module_info: Dict[nn.Parameter, ParamModuleInfo] = {}
+    # Do not remove duplicates to find parameters from shared modules
```

Comment:
Nit: I misunderstand and thought shared modules was referring to `Transformer` and `Block1` since they share `param` `Transformer(Block(param))`. But actually it's referring to `Block1` and `Block2` if Block1.param == Block2.param

maybe considering this docstring
```
\# Different modules may refer to the same parameter instance
\# Do not remove duplicates to find parameters shared by different modules
```